[SPARK-16217][SQL] Support SELECT INTO statement

wuxianxingkong opened the pull request:
## What changes were proposed in this pull request?

This PR implements the _SELECT INTO_ statement.

The _SELECT INTO_ statement selects data from one table and inserts it into a new table as follows.

```
SELECT column_name(s)
INTO newtable
FROM table1;
```

This statement is commonly used in SQL but not currently supported in SparkSQL.
We investigated the Catalyst and found that this statement can be implemented by improving the grammar and reusing the logical plan of _CTAS_.

The related JIRA is https://issues.apache.org/jira/browse/SPARK-16217
## How was this patch tested?

SQLQuerySuite.

wuxianxingkong committed:
SELECT INTO Implements

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -338,7 +338,7 @@ querySpecification
        fromClause?
        (WHERE where=booleanExpression)?)
-    | ((kind=SELECT setQuantifier? namedExpressionSeq fromClause?
+    | ((kind=SELECT setQuantifier? namedExpressionSeq (intoClause? fromClause)?
```
Hi, @wuxianxingkong .
Currently, the following seems to be not considered yet. Could you modify the syntax to support this too?
```
SELECT 1
INTO newtable
```

wuxianxingkong replied:
```sql
SELECT 1 
INTO newtable
```
This won't work because we need oldtable info to create newtable. So the sql should be
```sql
SELECT 1
INTO newtable 
FROM oldtable
```
The result from my test is: a new table called newtable was created, one column called 1 has the length of oldtable.rows.length and all elements are 1.
Did you mean there is no _FROM_?

dongjoon-hyun replied:
In the Spark Shell, please run the followings.
```
sql("select 1")
```

wuxianxingkong replied:
At first, I modify grammar:
![wrong select into](https://raw.githubusercontent.com/wuxianxingkong/storage/master/original_select_into.png)
But it will affect multiInsertQueryBody rule, i.e.:
```sql
FROM OLD_TABLE
INSERT INTO T1
SELECT C1
INSERT INTO T2
SELECT C2
```
The Syntax tree before adding intoClause is:
![right tree structure](https://raw.githubusercontent.com/wuxianxingkong/storage/master/original_muliinsert_tree.png)
After adding intoClause ,the tree will be:
![wrong tree structure](https://raw.githubusercontent.com/wuxianxingkong/storage/master/wrong_multiinserttree.png) This is because INSERT is a nonreserved keyword and matching strategy of antlr.
One of the ways I can think of is to change grammar like this:
![one way](https://raw.githubusercontent.com/wuxianxingkong/storage/master/solve_way.png)
This can solve the problem because antlr parser chooses the alternative specified first.
By the way, the grammar now can support "SELECT 1 INTO newtable" now.
But this will cause confusion about querySpecification rule because of the duplication. Is there any way to solve this problem?Thanks.

dongjoon-hyun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:
```diff
@@ -1755,4 +1755,97 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
+  test("select into(check relation)") {
+    val originalConf = sessionState.conf.convertCTAS
+
+    setConf(SQLConf.CONVERT_CTAS, true)
```
```
withSQLConf(SQLConf. CONVERT_CTAS.key -> "true") {
```

dongjoon-hyun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:
```diff
@@ -1755,4 +1755,97 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
+
+    val defaultDataSource = sessionState.conf.defaultDataSourceName
+    try {
+      sql("DROP TABLE IF EXISTS si1")
```
Please consider the following convention.
```
withTable("si1") {
```

dongjoon-hyun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:
```diff
@@ -1755,4 +1755,97 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
+  test("select into(check answer)") {
+    sql("DROP TABLE IF EXISTS si1")
+    sql("DROP TABLE IF EXISTS si2")
+    sql("DROP TABLE IF EXISTS si3")
```
```
withTable("si1", "si2", "si3") {
```

dongjoon-hyun commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:
```diff
@@ -1755,4 +1755,97 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
+    sql("SELECT 1 AS key,value INTO si3 FROM src LIMIT 1").collect()
+    intercept[AnalysisException] {
+      sql("SELECT key, value INTO si3 FROM src ORDER BY key, value").collect()
+    }
```
Checking the real error message is better.
```
val m = intercept[AnalysisException] {
     sql("SELECT key, value INTO si3 FROM src ORDER BY key, value").collect()
}.getMessage
assert(m.contains("your exception message"))
```

dongjoon-hyun commented:
Hi, @wuxianxingkong .
Although I'm just a contributor like you, I left a few comments for you because I like your PR.
I hope your PR will be merged soon.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -159,7 +159,9 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
       // Add organization statements.
       optionalMap(ctx.queryOrganization)(withQueryResultClauses).
       // Add insert.
-      optionalMap(ctx.insertInto())(withInsertInto)
```
This allows for the following syntax:
```sql
INSERT INTO tbl_a
SELECT *
INTO tbl_a
FROM tbl_b
```
Make sure that we cannot have both.

hvanhovell replied:
We also need to check what this does with multi-insert syntax, i.e.:
```sql
FROM tbl_a
INSERT INTO tbl_b
SELECT *
INSERT INTO tbl_c
SELECT *
INTO tbl_c
```

hvanhovell commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -347,6 +347,10 @@ querySpecification
     ;
 
+intoClause
+    : INTO tableIdentifier
```
It is easier to just put this in the `querySpecification` rule. Make sure you given the tableIdentifier a proper name

hvanhovell replied:
Could you also check what kind of a plan the following query produces:
```SQL
SELECT a INTO tbl_a FROM tbl_b
```
We might run into a weird syntax error here. If we do then we need to move the `INTO` keyword from the `nonReserved` rule to the `identifier` rule.

wuxianxingkong replied:
For example, the json data registed as table tbl_b is:
![json data](https://raw.githubusercontent.com/wuxianxingkong/storage/master/use_data.png)
@hvanhovell 
The Logical Plan of sql "SELECT a INTO tbl_a FROM tbl_b" is:
![json plan](https://raw.githubusercontent.com/wuxianxingkong/storage/master/print_json_plan.png)
The results match the expectations

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -1376,4 +1376,62 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
+   *   ...
+   * }}}
+   */
+  override protected def withSelectInto(
```
The code below is duplicates. Why are we not using the existing CTAS code path?

wuxianxingkong replied:
Reusing CTAS code path means we need to convert _IntoClauseContext_ to _CreateTableContext_ (or construct a new _CreateTableContext_),it might be difficult to archive. Maybe there is another way?

wuxianxingkong committed:
1.Support "SELECT 1 INTO newtable"
2.Add check in  multiinsertquery syntax:not allow multi insert and select into appear at the same time
3.Add check in singleinsertquery:not allow insert into and select into appear at the same time

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -395,6 +414,15 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
+  protected def withSelectInto(
+      ctx: IntoClauseContext,
+      query: LogicalPlan): LogicalPlan = withOrigin(ctx) {
+    throw new ParseException("Script Select Into is not supported", ctx)
```
Why throwing an `ParseException `?

gatorsmile commented:
@wuxianxingkong Are you still working on this? Thanks!

gatorsmile commented:
We are closing it due to inactivity. please do reopen if you want to push it forward. Thanks!
